// @gnosis-guild/zodiac/dist/cjs/contracts.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MasterCopyInitData = exports.ContractFactories = exports.ContractAbis = exports.ContractAddresses = exports.ContractVersions = exports.MasterCopyAddresses = exports.SupportedNetworks = void 0;
const Bridge_json_1 = __importDefault(require("./abi/Bridge.json"));
const CirculatingSupplyErc20_json_1 = __importDefault(require("./abi/CirculatingSupplyErc20.json"));
const CirculatingSupplyErc721_json_1 = __importDefault(require("./abi/CirculatingSupplyErc721.json"));
const Connext_json_1 = __importDefault(require("./abi/Connext.json"));
const Delay_json_1 = __importDefault(require("./abi/Delay.json"));
const Erc20Votes_json_1 = __importDefault(require("./abi/Erc20Votes.json"));
const Erc721Votes_json_1 = __importDefault(require("./abi/Erc721Votes.json"));
const ExitErc20_json_1 = __importDefault(require("./abi/ExitErc20.json"));
const ExitErc721_json_1 = __importDefault(require("./abi/ExitErc721.json"));
const MetaGuard_json_1 = __importDefault(require("./abi/MetaGuard.json"));
const ModuleProxyFactory_json_1 = __importDefault(require("./abi/ModuleProxyFactory.json"));
const MultisendEncoder_json_1 = __importDefault(require("./abi/MultisendEncoder.json"));
const OptimisticGovernor_json_1 = __importDefault(require("./abi/OptimisticGovernor.json"));
const OzGovernor_json_1 = __importDefault(require("./abi/OzGovernor.json"));
const Permissions_json_1 = __importDefault(require("./abi/Permissions.json"));
const RealityErc20_json_1 = __importDefault(require("./abi/RealityErc20.json"));
const RealityEth_json_1 = __importDefault(require("./abi/RealityEth.json"));
const RolesV1_json_1 = __importDefault(require("./abi/RolesV1.json"));
const RolesV2_json_1 = __importDefault(require("./abi/RolesV2.json"));
const ScopeGuard_json_1 = __importDefault(require("./abi/ScopeGuard.json"));
const Tellor_json_1 = __importDefault(require("./abi/Tellor.json"));
const types_1 = require("./factory/types");
const BridgeInitData = __importStar(require("./initData/Bridge"));
const CirculatingSupplyErc20InitData = __importStar(require("./initData/CirculatingSupplyErc20"));
const CirculatingSupplyErc721InitData = __importStar(require("./initData/CirculatingSupplyErc721"));
const ConnextInitData = __importStar(require("./initData/Connext"));
const DelayInitData = __importStar(require("./initData/Delay"));
const Erc20VotesInitData = __importStar(require("./initData/Erc20Votes"));
const Erc721VotesInitData = __importStar(require("./initData/Erc721Votes"));
const ExitErc20InitData = __importStar(require("./initData/ExitErc20"));
const ExitErc721InitData = __importStar(require("./initData/ExitErc721"));
const ModuleProxyFactoryInitData = __importStar(require("./initData/ModuleProxyFactory"));
const MultisendEncoderInitData = __importStar(require("./initData/MultisendEncoder"));
const OzGovernorInitData = __importStar(require("./initData/OzGovernor"));
const PermissionsInitData = __importStar(require("./initData/Permissions"));
const RealityErc20InitData = __importStar(require("./initData/RealityErc20"));
const RealityEthInitData = __importStar(require("./initData/RealityEth"));
const RolesV1InitData = __importStar(require("./initData/RolesV1"));
const RolesV2InitData = __importStar(require("./initData/RolesV2"));
const ScopeGuardInitData = __importStar(require("./initData/ScopeGuard"));
const types_2 = require("./types");
var SupportedNetworks;
(function (SupportedNetworks) {
    SupportedNetworks[SupportedNetworks["Mainnet"] = 1] = "Mainnet";
    SupportedNetworks[SupportedNetworks["GnosisChain"] = 100] = "GnosisChain";
    SupportedNetworks[SupportedNetworks["Goerli"] = 5] = "Goerli";
    SupportedNetworks[SupportedNetworks["ArbitrumOne"] = 42161] = "ArbitrumOne";
    SupportedNetworks[SupportedNetworks["Optimism"] = 10] = "Optimism";
    SupportedNetworks[SupportedNetworks["Polygon"] = 137] = "Polygon";
    SupportedNetworks[SupportedNetworks["ZkEvm"] = 1101] = "ZkEvm";
    SupportedNetworks[SupportedNetworks["Mumbai"] = 80001] = "Mumbai";
    SupportedNetworks[SupportedNetworks["Avalanche"] = 43114] = "Avalanche";
    SupportedNetworks[SupportedNetworks["BinanceSmartChain"] = 56] = "BinanceSmartChain";
    SupportedNetworks[SupportedNetworks["HardhatNetwork"] = 31337] = "HardhatNetwork";
    SupportedNetworks[SupportedNetworks["Linea"] = 59144] = "Linea";
    SupportedNetworks[SupportedNetworks["LineaGoerli"] = 59140] = "LineaGoerli";
    SupportedNetworks[SupportedNetworks["Sepolia"] = 11155111] = "Sepolia";
    SupportedNetworks[SupportedNetworks["CoreTestnet"] = 1115] = "CoreTestnet";
    SupportedNetworks[SupportedNetworks["Core"] = 1116] = "Core";
    SupportedNetworks[SupportedNetworks["Base"] = 8453] = "Base";
    SupportedNetworks[SupportedNetworks["BaseSepolia"] = 84532] = "BaseSepolia";
    SupportedNetworks[SupportedNetworks["Mantle"] = 5000] = "Mantle";
    SupportedNetworks[SupportedNetworks["Berachain"] = 80094] = "Berachain";
    SupportedNetworks[SupportedNetworks["Sonic"] = 146] = "Sonic";
    SupportedNetworks[SupportedNetworks["Celo"] = 42220] = "Celo";
    SupportedNetworks[SupportedNetworks["Bob"] = 60808] = "Bob";
    SupportedNetworks[SupportedNetworks["HyperEvm"] = 999] = "HyperEvm";
    SupportedNetworks[SupportedNetworks["WorldChain"] = 480] = "WorldChain";
    SupportedNetworks[SupportedNetworks["Flare"] = 14] = "Flare";
    SupportedNetworks[SupportedNetworks["Ink"] = 57073] = "Ink";
    SupportedNetworks[SupportedNetworks["Hemi"] = 43111] = "Hemi";
    SupportedNetworks[SupportedNetworks["Katana"] = 747474] = "Katana";
    SupportedNetworks[SupportedNetworks["Lens"] = 232] = "Lens";
    SupportedNetworks[SupportedNetworks["Peaq"] = 3338] = "Peaq";
    SupportedNetworks[SupportedNetworks["Unichain"] = 130] = "Unichain";
    SupportedNetworks[SupportedNetworks["ZkSync"] = 324] = "ZkSync";
    SupportedNetworks[SupportedNetworks["Scroll"] = 534352] = "Scroll";
    SupportedNetworks[SupportedNetworks["Aurora"] = 1313161554] = "Aurora";
    SupportedNetworks[SupportedNetworks["GnosisChiado"] = 10200] = "GnosisChiado";
})(SupportedNetworks || (exports.SupportedNetworks = SupportedNetworks = {}));
// const canonicalMasterCopyAddress = (contract: KnownContracts) => {
//   const { initCode, salt } = MasterCopyInitData[contract];
//   return getCreate2Address(SingletonFactoryAddress, salt, keccak256(initCode));
// };
/** The canonical mastercopy addresses when deployed using the singleton factory.  */
const CanonicalAddresses = {
    [types_1.KnownContracts.META_GUARD]: {
        "1.0.0": "0xe2847462a574bfd43014d1c7BB6De5769C294691",
    },
    [types_1.KnownContracts.REALITY_ETH]: {
        "2.0.0": "0x4e35DA39Fa5893a70A40Ce964F993d891E607cC0",
    },
    [types_1.KnownContracts.REALITY_ERC20]: {
        "2.0.0": "0x7276813b21623d89BA8984B225d5792943DD7dbF",
    },
    [types_1.KnownContracts.BRIDGE]: {
        "1.0.0": "0x03B5eBD2CB2e3339E93774A1Eb7c8634B8C393A9",
    },
    [types_1.KnownContracts.DELAY]: {
        "1.0.0": "0xD62129BF40CD1694b3d9D9847367783a1A4d5cB4",
        "1.0.1": "0xd54895B1121A2eE3f37b502F507631FA1331BED6",
        "1.1.0": "0x01F8cabB808D7dE0dF4202D4B60C8310d2f1339b",
    },
    [types_1.KnownContracts.FACTORY]: {
        "1.0.0": "0x00000000062c52e29e8029dc2413172f6d619d85",
        "1.1.0": "0x00000000000DC7F163742Eb4aBEf650037b1f588",
        "1.2.0": "0x000000000000aDdB49795b0f9bA5BC298cDda236",
    },
    [types_1.KnownContracts.EXIT_ERC20]: {
        "1.0.0": "0x35E35dcDc7Cd112B93C7c55987C86e5D6D419C69",
        "1.1.0": "0x33bCa41bda8A3983afbAd8fc8936Ce2Fb29121da",
        "1.2.0": "0x3ed380a282aDfA3460da28560ebEB2F6D967C9f5",
    },
    [types_1.KnownContracts.EXIT_ERC721]: {
        "1.1.0": "0xD3579C14a4181EfC3DF35C3103D20823A8C8d718",
        "1.2.0": "0xE0eCE32Eb4BE4E9224dcec6a4FcB335c1fe05CDe",
    },
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC20]: {
        "1.0.0": "0xd7a85e7D0813F8440602E243Acb67df3CCeb5a60",
        "1.1.0": "0xb50fab2e2892E3323A5300870C042B428B564FE3",
        "1.2.0": "0x5Ed57C291a184cc244F5c9B5E9F11a8DD08BBd12",
    },
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC721]: {
        "1.1.0": "0x71530ec830CBE363bab28F4EC52964a550C0AB1E",
        "1.2.0": "0xBD34D00dC0ae37C687F784A11FA6a0F2c5726Ba3",
    },
    [types_1.KnownContracts.SCOPE_GUARD]: {
        "1.0.0": "0xeF27fcd3965a866b22Fb2d7C689De9AB7e611f1F",
    },
    [types_1.KnownContracts.ROLES]: {
        "1.0.0": "0x85388a8cd772b19a468F982Dc264C238856939C9",
        "1.1.0": "0xD8DfC1d938D7D163C5231688341e9635E9011889",
    },
    [types_1.KnownContracts.ROLES_V1]: {
        "1.0.0": "0x85388a8cd772b19a468F982Dc264C238856939C9",
        "1.1.0": "0xD8DfC1d938D7D163C5231688341e9635E9011889",
    },
    [types_1.KnownContracts.PERMISSIONS]: {
        "1.0.0": "0x33D1C5A5B6a7f3885c7467e829aaa21698937597",
    },
    [types_1.KnownContracts.ROLES_V2]: {
        "2.1.0": "0x9646fDAD06d3e24444381f44362a3B0eB343D337",
    },
    [types_1.KnownContracts.TELLOR]: {
        "2.1.0": "0xa89EC2C1e218CfBb0F82829E95352CeAbDEe9A69",
    },
    [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
        "1.0.0": "",
    },
    [types_1.KnownContracts.OZ_GOVERNOR]: {
        "1.0.0": "0xe28c39FAC73cce2B33C4C003049e2F3AE43f77d5",
    },
    [types_1.KnownContracts.ERC20_VOTES]: {
        "1.0.0": "0x752c61de75ADA0F8a33e048d2F773f51172f033e",
    },
    [types_1.KnownContracts.ERC721_VOTES]: {
        "1.0.0": "0xeFf38b2eBB95ACBA09761246045743f40e762568",
    },
    [types_1.KnownContracts.MULTISEND_ENCODER]: {
        "1.0.0": "0xb67EDe523171325345780fA3016b7F5221293df0",
    },
    [types_1.KnownContracts.CONNEXT]: {
        "1.0.0": "0x7dE07b9De0bf0FABf31A188DE1527034b2aF36dB",
    },
};
/**
 * Canonical addresses of head versions of each contract.
 * This export will be removed in a future version. Use `ContractAddresses` instead.
 * @deprecated
 **/
exports.MasterCopyAddresses = Object.fromEntries(Object.entries(CanonicalAddresses).map(([key, value]) => [
    key,
    Object.values(value).pop() || "",
]));
/** Addresses of all deployed contracts in all versions */
exports.ContractVersions = {
    [SupportedNetworks.Mainnet]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x28CeBFE94a03DbCA9d17143e9d2Bd1155DC26D5d",
        }, [types_1.KnownContracts.TELLOR]: {
            "2.1.0": "0xaB80413Ab8C4aF91Af5F349352a0c92CF3A30D18",
        } }),
    [SupportedNetworks.Goerli]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x07a7Be7AA4AaD42696A17e974486cb64A4daC47b",
        }, [types_1.KnownContracts.TELLOR]: {
            "2.1.0": "0x4FCba614bf5df50b2575a5825adbc4739aE6FDF5",
        } }),
    [SupportedNetworks.BinanceSmartChain]: CanonicalAddresses,
    [SupportedNetworks.GnosisChain]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x972396Ab668cd11dc1F6321A5ae30c6A8d3759F0",
        } }),
    [SupportedNetworks.Polygon]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x3Cc4b597E9c3f51288c6Cd0c087DC14c3FfdD966",
        } }),
    [SupportedNetworks.ZkEvm]: CanonicalAddresses,
    [SupportedNetworks.HardhatNetwork]: CanonicalAddresses,
    [SupportedNetworks.Mumbai]: CanonicalAddresses,
    [SupportedNetworks.ArbitrumOne]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x30679ca4ea452d3df8a6c255a806e08810321763",
        } }),
    [SupportedNetworks.Optimism]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x357fe84E438B3150d2F68AB9167bdb8f881f3b9A",
        } }),
    [SupportedNetworks.Avalanche]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.TELLOR]: {}, [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0xEF8b46765ae805537053C59f826C3aD61924Db45",
        } }),
    [SupportedNetworks.Linea]: CanonicalAddresses,
    [SupportedNetworks.LineaGoerli]: CanonicalAddresses,
    [SupportedNetworks.Sepolia]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.TELLOR]: {
            "2.1.0": "0x5F05cd402BF4017970fe390D43B20C3dC3832CD9",
        }, [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x40153DdFAd90C49dbE3F5c9F96f2a5B25ec67461",
        } }),
    [SupportedNetworks.CoreTestnet]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0xD43463Fadd73373bE260b67F5825274F4403dAF0",
        } }),
    [SupportedNetworks.Core]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x596Fd6A5A185c67aBD1c845b39f593fBA9C233aa",
        } }),
    [SupportedNetworks.Base]: Object.assign(Object.assign({}, CanonicalAddresses), { [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: {
            "1.2.0": "0x80bCA2E1c272239AdFDCdc87779BC8Af6E12e633",
        } }),
    [SupportedNetworks.BaseSepolia]: CanonicalAddresses,
    [SupportedNetworks.Berachain]: CanonicalAddresses,
    [SupportedNetworks.Sonic]: CanonicalAddresses,
    [SupportedNetworks.Celo]: CanonicalAddresses,
    [SupportedNetworks.Mantle]: CanonicalAddresses,
    [SupportedNetworks.Bob]: CanonicalAddresses,
    [SupportedNetworks.HyperEvm]: CanonicalAddresses,
    [SupportedNetworks.Flare]: CanonicalAddresses,
    [SupportedNetworks.Ink]: CanonicalAddresses,
    [SupportedNetworks.WorldChain]: CanonicalAddresses,
    [SupportedNetworks.Hemi]: CanonicalAddresses,
    [SupportedNetworks.Katana]: CanonicalAddresses,
    [SupportedNetworks.Lens]: CanonicalAddresses,
    [SupportedNetworks.Peaq]: CanonicalAddresses,
    [SupportedNetworks.Unichain]: CanonicalAddresses,
    [SupportedNetworks.ZkSync]: CanonicalAddresses,
    [SupportedNetworks.Scroll]: CanonicalAddresses,
    [SupportedNetworks.Aurora]: CanonicalAddresses,
    [SupportedNetworks.GnosisChiado]: CanonicalAddresses,
};
/** Addresses of the head versions of all contracts */
exports.ContractAddresses = Object.fromEntries(Object.entries(exports.ContractVersions).map(([network, contracts]) => [
    network,
    Object.fromEntries(Object.entries(contracts).map(([contract, versions]) => [
        contract,
        Object.values(versions).pop() || "",
    ])),
]));
exports.ContractAbis = {
    [types_1.KnownContracts.BRIDGE]: Bridge_json_1.default,
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC20]: CirculatingSupplyErc20_json_1.default,
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC721]: CirculatingSupplyErc721_json_1.default,
    [types_1.KnownContracts.DELAY]: Delay_json_1.default,
    [types_1.KnownContracts.ERC20_VOTES]: Erc20Votes_json_1.default,
    [types_1.KnownContracts.ERC721_VOTES]: Erc721Votes_json_1.default,
    [types_1.KnownContracts.EXIT_ERC20]: ExitErc20_json_1.default,
    [types_1.KnownContracts.EXIT_ERC721]: ExitErc721_json_1.default,
    [types_1.KnownContracts.FACTORY]: ModuleProxyFactory_json_1.default,
    [types_1.KnownContracts.META_GUARD]: MetaGuard_json_1.default,
    [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: OptimisticGovernor_json_1.default,
    [types_1.KnownContracts.OZ_GOVERNOR]: OzGovernor_json_1.default,
    [types_1.KnownContracts.REALITY_ERC20]: RealityErc20_json_1.default,
    [types_1.KnownContracts.REALITY_ETH]: RealityEth_json_1.default,
    [types_1.KnownContracts.ROLES]: RolesV1_json_1.default,
    [types_1.KnownContracts.ROLES_V1]: RolesV1_json_1.default,
    [types_1.KnownContracts.ROLES_V2]: RolesV2_json_1.default,
    [types_1.KnownContracts.SCOPE_GUARD]: ScopeGuard_json_1.default,
    [types_1.KnownContracts.TELLOR]: Tellor_json_1.default,
    [types_1.KnownContracts.MULTISEND_ENCODER]: MultisendEncoder_json_1.default,
    [types_1.KnownContracts.PERMISSIONS]: Permissions_json_1.default,
    [types_1.KnownContracts.CONNEXT]: Connext_json_1.default,
};
exports.ContractFactories = {
    [types_1.KnownContracts.BRIDGE]: types_2.factories.Bridge__factory,
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC20]: types_2.factories.CirculatingSupplyErc20__factory,
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC721]: types_2.factories.CirculatingSupplyErc721__factory,
    [types_1.KnownContracts.DELAY]: types_2.factories.Delay__factory,
    [types_1.KnownContracts.ERC20_VOTES]: types_2.factories.Erc20Votes__factory,
    [types_1.KnownContracts.ERC721_VOTES]: types_2.factories.Erc721Votes__factory,
    [types_1.KnownContracts.EXIT_ERC20]: types_2.factories.ExitErc20__factory,
    [types_1.KnownContracts.EXIT_ERC721]: types_2.factories.ExitErc721__factory,
    [types_1.KnownContracts.FACTORY]: types_2.factories.ModuleProxyFactory__factory,
    [types_1.KnownContracts.META_GUARD]: types_2.factories.MetaGuard__factory,
    [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: types_2.factories.OptimisticGovernor__factory,
    [types_1.KnownContracts.OZ_GOVERNOR]: types_2.factories.OzGovernor__factory,
    [types_1.KnownContracts.REALITY_ERC20]: types_2.factories.RealityErc20__factory,
    [types_1.KnownContracts.REALITY_ETH]: types_2.factories.RealityEth__factory,
    [types_1.KnownContracts.ROLES]: types_2.factories.RolesV1__factory,
    [types_1.KnownContracts.ROLES_V1]: types_2.factories.RolesV1__factory,
    [types_1.KnownContracts.ROLES_V2]: types_2.factories.RolesV2__factory,
    [types_1.KnownContracts.SCOPE_GUARD]: types_2.factories.ScopeGuard__factory,
    [types_1.KnownContracts.TELLOR]: types_2.factories.Tellor__factory,
    [types_1.KnownContracts.MULTISEND_ENCODER]: types_2.factories.MultisendEncoder__factory,
    [types_1.KnownContracts.PERMISSIONS]: types_2.factories.Permissions__factory,
    [types_1.KnownContracts.CONNEXT]: types_2.factories.Connext__factory,
};
exports.MasterCopyInitData = {
    [types_1.KnownContracts.BRIDGE]: BridgeInitData,
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC20]: CirculatingSupplyErc20InitData,
    [types_1.KnownContracts.CIRCULATING_SUPPLY_ERC721]: CirculatingSupplyErc721InitData,
    [types_1.KnownContracts.DELAY]: DelayInitData,
    [types_1.KnownContracts.ERC20_VOTES]: Erc20VotesInitData,
    [types_1.KnownContracts.ERC721_VOTES]: Erc721VotesInitData,
    [types_1.KnownContracts.EXIT_ERC20]: ExitErc20InitData,
    [types_1.KnownContracts.EXIT_ERC721]: ExitErc721InitData,
    [types_1.KnownContracts.FACTORY]: ModuleProxyFactoryInitData,
    [types_1.KnownContracts.META_GUARD]: undefined,
    [types_1.KnownContracts.OPTIMISTIC_GOVERNOR]: undefined,
    [types_1.KnownContracts.OZ_GOVERNOR]: OzGovernorInitData,
    [types_1.KnownContracts.REALITY_ERC20]: RealityErc20InitData,
    [types_1.KnownContracts.REALITY_ETH]: RealityEthInitData,
    [types_1.KnownContracts.ROLES]: RolesV1InitData,
    [types_1.KnownContracts.ROLES_V1]: RolesV1InitData,
    [types_1.KnownContracts.ROLES_V2]: RolesV2InitData,
    [types_1.KnownContracts.SCOPE_GUARD]: ScopeGuardInitData,
    [types_1.KnownContracts.TELLOR]: undefined,
    [types_1.KnownContracts.MULTISEND_ENCODER]: MultisendEncoderInitData,
    [types_1.KnownContracts.PERMISSIONS]: PermissionsInitData,
    [types_1.KnownContracts.CONNEXT]: ConnextInitData,
};
