// @torchauth/vite-plugin-wrangler-spa/dist/index.js
import { unstable_dev } from 'wrangler';
import { builtinModules } from 'node:module';
import { transform } from '@swc/core';
import { writeFile } from 'node:fs';

const miniflarePlugin = (config) => {
    const { functionEntrypoint, wranglerConfig, allowedApiPaths } = config;
    // force wrangler settings that are required for function HMR to work
    if (!wranglerConfig.experimental)
        wranglerConfig.experimental = {};
    wranglerConfig.experimental.liveReload = true;
    wranglerConfig.experimental.testMode = false;
    return {
        name: 'vite-plugin-wrangler-spa:miniflare',
        apply: 'serve',
        configResolved: async (viteConfig) => {
            viteConfig.server.proxy = allowedApiPaths.reduce((acc, curr) => ({ ...acc, [curr]: `http://127.0.0.1:${wranglerConfig.port}` }), {});
        },
        configureServer: async (devServer) => {
            return async () => {
                const wranglerDevServer = await unstable_dev(functionEntrypoint, wranglerConfig);
                devServer.httpServer?.on('close', async () => await wranglerDevServer.stop());
            };
        },
        handleHotUpdate: async (ctx) => {
            if (ctx.file.includes(functionEntrypoint.split('/')[0]))
                ctx.server.hot.send({
                    type: 'custom',
                    event: 'function-update',
                    data: { file: ctx.file },
                });
        },
        transformIndexHtml: {
            order: 'pre',
            handler: () => [
                {
                    tag: 'script',
                    attrs: { type: 'module' },
                    children: browserHmrNotification,
                },
            ],
        },
    };
};
const browserHmrNotification = `
if (import.meta.hot) {
  let outputColor = "color:cyan; font-weight:bold;"
  import.meta.hot.on('function-update', (data) => {
    console.log(\`%c 🔥 Cloudflare Pages Function - update detected in file: '\${data.file}'\`, outputColor);
    location.reload(true);
  }); 
}`;

const getViteConfig = ({ functionEntrypoint, external }) => {
    return {
        ssr: {
            external,
            noExternal: true,
        },
        esbuild: false, // we use SWC to build
        build: {
            sourcemap: true, // always include sourcemaps
            rollupOptions: {
                external: [...builtinModules, /^node:/],
                input: functionEntrypoint,
                preserveEntrySignatures: 'allow-extension',
                output: { entryFileNames: '_worker.js' },
            },
            emptyOutDir: false,
            ssr: true,
        },
    };
};

const swcPlugin = (config) => {
    const { allowedApiPaths, excludedApiPaths, swcConfig } = config;
    return {
        name: 'vite-plugin-wrangler-spa:swc',
        apply: (_, { command, mode }) => command === 'build' && mode === 'page-function',
        config: () => getViteConfig(config),
        transform: (code) => transform(code, swcConfig),
        writeBundle: async () => await writeFile('dist/_routes.json', JSON.stringify({
            version: 1,
            include: allowedApiPaths.map((x) => x.replace('^', '')),
            exclude: excludedApiPaths.map((x) => x.replace('^', '')),
        }, null, 2), (err) => (err ? console.error(err.message) : null)),
    };
};

const defaultCloudflareSpaConfig = {
    allowedApiPaths: ['^/api/*'],
    excludedApiPaths: [],
    functionEntrypoint: 'functions/index.ts',
    external: [],
    wranglerConfig: {
        port: 55554,
        logLevel: 'log',
        experimental: {
            disableExperimentalWarning: true, //disable because it's annoying
        },
    },
    wranglerConfigPath: 'wrangler.toml',
    swcConfig: {
        minify: true,
        sourceMaps: true,
        inlineSourcesContent: true,
        jsc: {
            target: 'esnext',
            parser: {
                tsx: true,
                syntax: 'typescript',
                decorators: true,
            },
            transform: {
                decoratorMetadata: true,
                legacyDecorator: true,
            },
            minify: {
                compress: true,
                mangle: true,
            },
            loose: true,
        },
    },
};
const viteWranglerSpa = (config = defaultCloudflareSpaConfig) => {
    const runtimeConfig = {
        ...defaultCloudflareSpaConfig,
        ...config,
    };
    return [miniflarePlugin(runtimeConfig), swcPlugin(runtimeConfig)];
};

export { viteWranglerSpa };
